### PR TITLE
feat: add typescript

### PR DESCRIPTION
**Pull request by KrzysztofMoch**
## Summary
This PR add built-in types and configuration for typescript. It is also step to prepare for new architecture support (codegen). Changes and types are based on [New Arch PR](https://github.com/react-native-video/react-native-video/pull/3204) by @yungblud 

## Why is needed ?
With built-in types we have up to date types on every version and we don't need to install an rely on additional package for types `@types/react-native-video`. Overall built-in types **increase the comfort of both developers and maintainers**.
But `@types/react-native-video` **should stay** to support older versions of `react-native-video`

## How to test ?
1. Checkout for branch of PR
2. Install dependencies
3. Run `yarn build`
4. update dependencies in (basic) example app
5. Check if types are correctly imported
6. Example app should build and run
7. Types for `Video` component should be available

### CODE REVIEW

File: src/types/video.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+type SelectedTrack = {
+  type: 'system' | 'disabled' | 'title' | 'language' | 'index';
+  value?: string | number;
+}
+
+type SelectedVideoTrack = {
+  type: 'auto' | 'disabled' | 'resolution' | 'index'
+  value: number;
```
Comment (freeboub) [inline src/types/video.ts]: i think value is not mandatory here ? For auto configuration. And can it be a string for resolution?

File: src/VideoNativeComponent.ts [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+  onVideoExternalPlaybackChange?: (event: NativeSyntheticEvent<OnExternalPlaybackChangeData>) => void;
+  onGetLicense?: (event: NativeSyntheticEvent<OnGetLicenseData>) => void;
+  onPictureInPictureStatusChanged?: (event: NativeSyntheticEvent<OnPictureInPictureStatusChangedData>) => void;
+  onRestoreUserInterfaceForPictureInPictureStop?: (event: NativeSyntheticEvent<Readonly<{}>>) => void;
+  onReceiveAdEvent?: (event: NativeSyntheticEvent<OnReceiveAdEventData>) => void;
+  onVideoPlaybackStateChanged?: (event: NativeSyntheticEvent<OnPlaybackStateChangedData>) => void; // android only
+  onVideoIdle?: (event: NativeSyntheticEvent<{}>) => void; // android only (nowhere in document, so do not use as props. just type declaration)
+  onAudioFocusChanged?: (event: NativeSyntheticEvent<OnAudioFocusChangedData>) => void; // android only (nowhere in document, so do not use as props. just type declaration)
+  onTimedMetadata?: (event: NativeSyntheticEvent<OnTimedMetadataData>) => void; // ios, android
+  onAudioTracks: (event: NativeSyntheticEvent<OnAudioTracksData>) => void; // android
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Missing optionnal arg ?

File: src/VideoNativeComponent.ts [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+  error: string;
+}>
+
+export type OnAudioFocusChangedData = Readonly<{
+  hasFocus: boolean;
+}>
+
+export type NativeVideoResizeMode = 'ScaleNone' | 'ScaleToFill' | 'ScaleAspectFit' | 'ScaleAspectFill';
+export interface VideoNativeProps extends ViewProps {
+  src: VideoSrc;
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Missing '?'
Client app can put src=undefined to stop the player

File: src/VideoNativeComponent.ts [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+     * iOS only supports VTT, Android supports all 3
+     */
+    type?: 'srt' | 'ttml' | 'vtt'
+    selected?: boolean
+  }>>
+}>
+
+export type OnVideoTracksData = Readonly<{
+  videoTracks: ReadonlyArray<Readonly<{
+    trackId?: number
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Track id is mandatory I think you can remove the ?

File: src/VideoNativeComponent.ts [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    language?: string
+    bitrate?: number
+    type?: string
+    selected?: boolean
+  }>>
+}>
+
+export type OnTextTracksData = Readonly<{
+  textTracks: ReadonlyArray<Readonly<{
+    index?: number
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Index is mandatory you can remove ?

File: src/VideoNativeComponent.ts [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+  metadata: ReadonlyArray<Readonly<{
+    value?: string
+    identifier?: string
+  }>>
+}>
+
+
+export type OnAudioTracksData = Readonly<{
+  audioTracks: ReadonlyArray<Readonly<{
+    index?: number
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Index is mandatory you can remove the ?

File: src/VideoNativeComponent.ts [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+}>
+
+export type OnPlaybackStateChangedData = Readonly<{
+  isPlaying: boolean;
+}>
+
+export type OnTimedMetadataData = Readonly<{
+  metadata: ReadonlyArray<Readonly<{
+    value?: string
+    identifier?: string
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Not sure but identifier is mandatory?! Can you double check please ?

File: src/VideoNativeComponent.ts [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+  paddingRight?: number;
+}>
+
+export type OnLoadData = Readonly<{
+  currentTime: number;
+  duration: number;
+  naturalSize: Readonly<{
+    width: number;
+    height: number;
+    orientation: string;
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: There is no enum for orientation? That is sad 😢

File: src/VideoNativeComponent.ts [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+  maxBufferMs?: number;
+  bufferForPlaybackMs?: number;
+  bufferForPlaybackAfterRebufferMs?: number;
+  maxHeapAllocationPercent?: number;
+  minBackBufferMemoryReservePercent?: number;
+  minBufferMemoryReservePercent?: number;
+}>
+
+type SelectedVideoTrack = Readonly<{
+  type?: 'auto' | 'disabled' | 'resolution' | 'index';
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Type is mandatory i think

File: src/VideoNativeComponent.ts [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+type SelectedTextTrack = Readonly<{
+  selectedTextType?: TextTrackType;
+  value?: string;
+  index?: number;
+}>
+
+type AudioTrackType = 'system' | 'disabled' | 'title' | 'language' | 'index';
+
+type SelectedAudioTrack = Readonly<{
+  selectedAudioType?: AudioTrackType;
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: SelectedAudioType is mandatory?!

File: src/VideoNativeComponent.ts [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+  title: string;
+  language: string;
+  type: string;
+  uri: string;
+}>>
+
+type TextTrackType = 'system' | 'disabled' | 'title' | 'language' | 'index';
+
+type SelectedTextTrack = Readonly<{
+  selectedTextType?: TextTrackType;
```
Comment (freeboub) [inline src/VideoNativeComponent.ts]: Selected texttype is mandatory?!

**Review by freeboub — CHANGES_REQUESTED** [review]

**Comment by freeboub** [comment]
Great 👍 a lot of minor comments but looks good

File: src/types/video.ts (hunk 1 above)
Comment (KrzysztofMoch) [reply src/types/video.ts]: Yes value is not mandatory but resolution is number

File: src/VideoNativeComponent.ts (hunk 2 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

File: src/VideoNativeComponent.ts (hunk 3 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

File: src/VideoNativeComponent.ts (hunk 4 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

File: src/VideoNativeComponent.ts (hunk 5 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

File: src/VideoNativeComponent.ts (hunk 6 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

File: src/VideoNativeComponent.ts (hunk 7 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Should be mandatory, Fixed ✅

File: src/VideoNativeComponent.ts (hunk 8 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Added `'portrait' | 'landscape'` for now

File: src/VideoNativeComponent.ts (hunk 9 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

File: src/VideoNativeComponent.ts (hunk 10 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

File: src/VideoNativeComponent.ts (hunk 11 above)
Comment (KrzysztofMoch) [reply src/VideoNativeComponent.ts]: Fixed ✅

**Comment by freeboub** [comment]
sorry @KrzysztofMoch I merge the DRM PR, so there are conflict with this one ...

**Review by freeboub — APPROVED** [review]

**Pull request merged**